### PR TITLE
Make Array differentiation work with non sympy expression

### LINKED ISSUE #20655: Array differentiation of not sympy expression
I noticed that the function ```derive_by_array``` doesn't accept as argument an integer or a float.
```python
import sympy as sp
from sympy.abc import x, y, z

U = sp.Array([x, y, z])
E = 2
G = sp.derive_by_array(E, U)
```

This code gives the following error:
```
AttributeError: 'int' object has no attribute 'diff'
```

It could be solved using a verification in the begin of the function:
```python
if isinstance(E, (int, float)):
    return sp.Array([0 for i in range(len(U))])
```

### PR DESCRIPTION
**Pull request by sidhu1012**
#### References to other Issues or PRs

Fixes #20655 

#### Brief description of what is fixed or changed
Make derive_by_array sympify it's arguments first.

#### Release Notes

* tensor
    * Make Array differentiation(`derive_by_array`) work with non sympy expressions.

### CODE REVIEW

**Comment by sylee957** [comment]
Why not use _sympify instead of sympify?

**Comment by sidhu1012** [comment]
> Why not use _sympify instead of sympify?

Because it caused some tests to fail with `SympifyError`.

**Comment by oscarbenjamin** [comment]
Which tests?

**Comment by sidhu1012** [comment]
> Which tests?

All these tests-

https://github.com/sympy/sympy/blob/5288eb336d12e4f3c3070bce67d75f38ab4d8376/sympy/tensor/array/tests/test_arrayop.py#L78-L95

**Comment by sidhu1012** [comment]
@sylee957 @oscarbenjamin Is `sympify` good then?

**Comment by oscarbenjamin** [comment]
Perhaps the arguments should be convert to Array using `Array(arg)` or something but we should not use `sympify`.

**Comment by sidhu1012** [comment]
> Perhaps the arguments should be convert to Array using `Array(arg)` or something but we should not use `sympify`.

But why shouldn't we use `sympify` here?

**Comment by asmeurer** [comment]
_sympify is better than sympify because it disallows strings, which is preferable for library functions. But it seems it also disallows lists, which we actually want to allow here. Actually sympify() converts lists to list, but we really want to convert it to an Array. If arrays are the only thing that make sense in this function, it may be better to use Array() instead of sympify() here.

File: sympy/tensor/array/arrayop.py [hunk 1]
```diff
@@ -269,6 +269,9 @@ def derive_by_array(expr, dx):
     from sympy.tensor.array import SparseNDimArray
     array_types = (Iterable, MatrixBase, NDimArray)
 
+    expr = sympify(expr)
+    dx = sympify(dx)
+
     if isinstance(dx, array_types):
```
Comment (asmeurer) [inline sympy/tensor/array/arrayop.py:275]: Can we just replace this `if` with a `try` that calls the constructor and sees if it works.

Comment (asmeurer) [reply sympy/tensor/array/arrayop.py:275]: Or use `_sympify` but just move it below these `if` blocks.

Comment (sidhu1012) [reply sympy/tensor/array/arrayop.py]: > Or use `_sympify` but just move it below these `if` blocks.

Yes it worked. Thanks!!

**Comment by sidhu1012** [comment]
@asmeurer @oscarbenjamin @sylee957 Please review!

**Pull request merged**